fish_prompt.fish

```diff
@@ -4,11 +4,170 @@
+    set git_branch[3] (__parse_git_branch)
+
+    # Check if there are files to commit
+    set -l is_git_dirty (command git status --porcelain --ignore-submodules ^/dev/null)
+
+    if test -n "$is_git_dirty"
+      set git_dirty[3] "$pure_symbol_git_dirty"
+    end
+
+    # Style Git branch output
+    set git_branch[2] "$pure_color_gray"
+    # Prefer space before Git branch name
+    set git_branch[1] " "
+
+    # Check if there is an upstream configured
+    command git rev-parse --abbrev-ref '@{upstream}' >/dev/null ^&1; and set -l has_upstream
+    if set -q has_upstream
+      command git rev-list --left-right --count 'HEAD...@{upstream}' | read -la git_status
+
+      set -l git_arrow_left $git_status[1]
+      set -l git_arrow_right $git_status[2] 
+
+      # If arrow is not "0", it means it's dirty
+      if test $git_arrow_left != 0
+        set git_arrows[3] "$pure_symbol_git_up_arrow"
+      end
+
+      if test $git_arrow_right != 0
+        set git_arrows[3] "$git_arrows[3]$pure_symbol_git_down_arrow"
+      end
+
+      # Format Git arrows
+      set git_arrows[2] "$pure_color_cyan"
+      set git_arrows[1] " "
+    end
+  end
+
+  # Get command execution duration
+  if test -n "$CMD_DURATION"
+    set command_duration[3] (__format_time $CMD_DURATION $pure_command_max_exec_time)
+  end
+  # Format command execution duration
+  set command_duration[2] "$pure_color_yellow"
+
+  # List elements so far
+  set -l current_elements \
+    $user $at $host \
+    $git_branch $git_dirty $git_arrows \
+    $command_duration
+
+  # Compute current length of the elements
+  set -l current_pre_prompt_length 0
+  for i in (seq (math (count $current_elements) / 3))
+    # Get preferred delimeter
+    set -l delimiter $current_elements[(math "3 * $i - 2")]
+    # Get textual representation
+    set -l text $current_elements[(math "3 * $i")]
+    # Ignore empty elements
+    if test -n "$text"
+      # Compute number of symbols
+      set current_pre_prompt_length (math $current_pre_prompt_length + (string length "$delimiter$text"))
+    end
+  end
+  # Substract number of symbols in first delimeter (depends on relative position of elements, should be revised if we'll have configuration that breaks it)
+  set current_pre_prompt_length (math $current_pre_prompt_length - (string length "$current_elements[$1]"))
+
+  # Format current folder on prompt output
+  set current_folder[3] (__parse_current_folder (math $COLUMNS - $current_pre_prompt_length))
+  set current_folder[2] "$pure_color_blue"
+  set current_folder[1] " "
+
+  set -l elements ""
+  # Define relative position of elements
+  if test $pure_user_host_location -eq 0
+    set elements \
+      $current_folder \
+      $git_branch $git_dirty $git_arrows \
+      $user $at $host \
+      $command_duration
+  else
+    set elements \
+      $user $at $host \
+      $current_folder \
+      $git_branch $git_dirty $git_arrows \
+      $command_duration
+  end
+
+  set -l pre_prompt ""
```

Comment:
Can be extracted to `pre_prompt.fish` (yes no mistake, it's related to previous) and call `get_pre_prompt`.